### PR TITLE
deps(ui): Upgrade to typescript 4.6

### PR DESCRIPTION
**Pull request by scttcper**
https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/

### CODE REVIEW

File: static/app/components/searchSyntax/parser.tsx [hunk 1]
```diff
@@ -529,9 +529,12 @@ export class TokenConverter {
    *
    * [0]:https://pegjs.org/documentation
    */
-  predicateFilter = <T extends FilterType>(type: T, key: FilterMap[T]['key']) => {
+  predicateFilter = <T extends FilterType>(
+    type: T,
+    key: ReturnType<TokenConverter['tokenKeyAggregate']>
```
Comment (scttcper) [inline static/app/components/searchSyntax/parser.tsx:534]: @evanpurkhiser this felt appropriate

File: static/app/components/searchSyntax/parser.tsx [hunk 2]
```diff
@@ -530,6 +530,7 @@ export class TokenConverter {
    * [0]:https://pegjs.org/documentation
    */
   predicateFilter = <T extends FilterType>(type: T, key: FilterMap[T]['key']) => {
+    // @ts-expect-error Unclear why this isn’t resolving correctly
```
Comment (JonasBa) [inline static/app/components/searchSyntax/parser.tsx:533]: I think this is related to https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/#indexed-access-inference-improvements, it probably does work better now :D

**Review by JonasBa — APPROVED** [review]
I would spend a bit of time to try and see if you can fix the ts-expect error, but otherwise LGTM

**Review by priscilawebdev — APPROVED** [review]

File: static/app/components/searchSyntax/parser.tsx (hunk 2 above)
Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:533]: @JonasBa I looked at this for a few hours, if you want to try and understand why it's not resolving correctly be my guest 😬

**Pull request merged**